pkg/objstore/oss/oss.go

```diff
@@ -0,0 +1,308 @@
+			"and delete it after test. Unset OSS_BUCKET env variable to use default logic. If you really want to run " +
+			"tests against provided (NOT USED!) bucket, set THANOS_ALLOW_EXISTING_BUCKET_USE=true.")
+	}
+
+	return NewTestBucketFromConfig(t, location, c, true)
+}
+
+// Upload the contents of the reader as an object into the bucket.
+func (b *Bucket) Upload(ctx context.Context, name string, r io.Reader) error {
+	if err := b.bucket.PutObject(name, r); err != nil {
+		return errors.Wrap(err, "upload oss object")
+	}
+	return nil
+}
+
+// Delete removes the object with the given name.
+func (b *Bucket) Delete(ctx context.Context, name string) error {
+	if err := b.bucket.DeleteObject(name); err != nil {
+		return errors.Wrap(err, "delete oss object")
+	}
+	return nil
+}
+
+func configFromEnv() Config {
+	c := Config{
+		EndPoint:        os.Getenv("OSS_ENDPOINT"),
+		Bucket:          os.Getenv("OSS_BUCKET"),
+		AccessKeyId:     os.Getenv("OSS_ACCESS_KEY_ID"),
+		AccessKeySecret: os.Getenv("OSS_ACCESS_KEY_SECRET"),
+	}
+	return c
+}
+
+// ValidateForTests checks to see the config options for tests are set.
+func ValidateForTests(conf Config) error {
+	if conf.EndPoint == "" ||
+		conf.AccessKeyId == "" ||
+		conf.AccessKeySecret == "" {
+		return errors.New("insufficient oss test configuration information")
+	}
+	return nil
+}
+
+func parseConfig(conf []byte) (Config, error) {
+	var config Config
+	if err := yaml.Unmarshal(conf, &config); err != nil {
+		return Config{}, err
+	}
+
+	return config, nil
+}
+
+func NewBucket(logger log.Logger, conf []byte, component string) (*Bucket, error) {
+	config, err := parseConfig(conf)
+	if err != nil {
+		return nil, err
+	}
+	if err := validate(config); err != nil {
+		return nil, err
+	}
+
+	client, err := alioss.New(config.EndPoint, config.AccessKeyId, config.AccessKeySecret)
+	if err != nil {
+		return nil, err
+	}
+	bk, err := client.Bucket(config.Bucket)
+	if err != nil {
+		return nil, errors.Wrap(err, "oss bucket")
+	}
+
+	bkt := &Bucket{
+		logger: logger,
+		client: client,
+		name:   config.Bucket,
+		config: config,
+		bucket: bk,
+	}
+	return bkt, nil
+}
+
+// Iter calls f for each entry in the given directory (not recursive.). The argument to f is the full
+// object name including the prefix of the inspected directory.
+func (b *Bucket) Iter(ctx context.Context, dir string, f func(string) error) error {
+	if dir != "" {
+		dir = strings.TrimSuffix(dir, objstore.DirDelim) + objstore.DirDelim
+	}
+
+	marker := alioss.Marker("")
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+		}
+		objects, err2 := b.bucket.ListObjects(alioss.Prefix(dir), alioss.Delimiter(objstore.DirDelim), marker)
+		if err2 != nil {
+			return errors.Wrap(err2, "oss bucket list")
+		}
+		marker = alioss.Marker(objects.NextMarker)
+
+		for _, object := range objects.Objects {
+			if object.Key == "" {
```

Comment:
Please add a comment why is this needed.

Does it skip directories? Or files with empty names?

Reply:
It does not need the condition `object.Key == ""` . For the first version of oss.go , I do not use `Marker` , so  I need to jump over  the root dir naming "" . THX ~~ I will correct it .